clang/test/Analysis/store-to-immutable-basic.cpp

```diff
@@ -0,0 +1,166 @@
+}
+
+// Write to const array parameter
+void test_write_to_const_array_param(const int arr[5]) {
+  *(int*)arr = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+// Write to const struct parameter
+struct ParamStruct {
+  const int z; // expected-note 3{{Memory region is in immutable space}}
+  int w;
+};
+
+void test_write_to_const_struct_param(const ParamStruct s) {
+  *(int*)&s.z = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+// Write to const struct reference parameter
+void test_write_to_const_struct_ref_param(const ParamStruct &s) {
+  *(int*)&s.z = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+// Write to const struct pointer parameter
+void test_write_to_const_struct_ptr_param(const ParamStruct *s) {
+  *(int*)&s->z = 100; // expected-warning {{Writing to immutable memory is undefined behavior}}
+  // expected-note@-1 {{Writing to immutable memory is undefined behavior. This memory region is marked as immutable and should not be modified}}
+}
+
+//===--- NEGATIVE TEST CASES ---===//
+// These tests should NOT trigger warnings
+
+// Write to non-const variable (should not warn)
+void test_write_to_nonconst() {
+  int non_const = 42;
+  *(int*)&non_const = 100; // No warning expected
+}
+
+// Write to non-const global variable (should not warn)
+int global_non_const = 42;
+
+void test_write_to_nonconst_global() {
+  *(int*)&global_non_const = 100; // No warning expected
+}
+
+// Write to non-const struct member (should not warn)
+struct NonConstStruct {
+  int x;
+  int y;
+};
+
+void test_write_to_nonconst_struct_member() {
+  NonConstStruct s = {1, 2};
+  *(int*)&s.x = 100; // No warning expected
+}
+
+// Write to non-const parameter (should not warn)
+void test_write_to_nonconst_param(int param) {
+  *(int*)&param = 100; // No warning expected
+}
+
+// Normal assignment to non-const variable (should not warn)
+void test_normal_assignment() {
+  int x = 42;
+  x = 100; // No warning expected
+}
+
+// Write to non-const data through const pointer (should not warn - underlying memory is non-const)
+void test_const_ptr_to_nonconst_data() {
+  int data = 42;
+  const int *ptr = &data;
+  *(int*)ptr = 100; // No warning expected
+}
+
+// Write to non-const data through const reference (should not warn - underlying memory is non-const)
+void test_const_ref_to_nonconst_data() {
+  int data = 42;
+  const int &ref = data;
+  *(int*)&ref = 100; // No warning expected
+} 
```

Comment:
Can you test the cases when `data` is actually `const`?

Reply:
I believe that test case would be the one at the top of the file:
```
void test_write_through_const_pointer() {
  const int local_const = 10;
  int *ptr = (int*)&local_const;
  *ptr = 20; // expected-warning {{Writing to immutable memory is undefined behavior}}
}
```

Reply:
It's not quite what I was looking for - my bad.
You are right, that `test_write_through_const_pointer` case already tested that case.
I was curious about the same but using references:

```c++
void test_ref_to_nonconst_data() {
  int data = 42;
  int &ref = data;
  ref = 100; // No warning expected
}

void test_ref_to_const_data() {
  const int data = 42;
  int &ref = *(int*)&data;
  ref = 100; // warn: ...
}
```

Reply:
Nice, I have added these cases.

Reply:
I think we still could not reach consensus here, and no test cases fulfill my request completely.
All writes in the tests happen through a pointer expression. Usually by taking the address of the reference then casting it to a wrong pointer type which is then dereferenced.

However, what I point to in my example is that forming a wrong reference is valid, as per C++ standard.
So this code is fine:
```
void test_ref_to_const_data() {
  const int data = 42;
  int &ref = *(int*)&data; // no UB
  // we are allowed to not use `ref`.
}
```
But then the `ref = 100` should trigger the checkBind and the checker diagnosing the issue.

Reply:
Ah, I just noticed that I might have reacted too soon, as the PR was not updated, so your proposed changes were not pushed yet. nvm